refactor(configs): add configs for deployments to environments (#3265)
juspay/hyperswitch: config/deployments/README.md

# Configs for deployments

## Introduction

> This directory contains the configs for deployments of hyperswitch in different hosted environments.

Reviewer: Please ensure you follow similar case convention throughout the document: either `Hyperswitch` or `hyperswitch`.

Hyperswitch has **3** components namely,

- router
- drainer
- scheduler
  - consumer
  - producer

> Hyperswitch maintains 3 different environments for deployments under the `router` component, namely,

Reviewer: ```suggestion
We maintain configs for the `router` component for 3 different environments, namely,
```

- Integration Test
- Sandbox
- Production

To learn about what "router", "drainer" and "scheduler" is, please refer to the [Hyperswitch architecture][architecture] documentation.

### Tree structure

```text
config/deployments      # Root directory for the deployment configs
├── README.md           # This file
├── drainer.toml        # Config specific to drainer
├── env_specific.toml   # Config for environment specific values which are meant to be sensitive (to be set by the user)

> ├── integ.toml          # Config specific to integration_test environment

Reviewer: ```suggestion
├── integration_test.toml          # Config specific to integration_test environment
```

├── production.toml     # Config specific to production environment
├── sandbox.toml        # Config specific to sandbox environment
└── scheduler           # Directory for scheduler configs
    ├── consumer.toml   # Config specific to consumer
    └── producer.toml   # Config specific to producer
```

## Router

The `integration_test.toml`, `sandbox.toml`, and `production.toml` files are configuration files for the environments `integration_test`, `sandbox`, and `production`, respectively. These files maintain a 1:1 mapping with the environment names, and it is recommended to use the same name for the environment throughout this document.

### Generating a Config File for the Router

> The `env_specific.toml` file contains values that are specific to the environment. This file is kept separate because the values in it are sensitive and are meant to be set by the user. The `env_specific.toml` file is merged with the `integ.toml`, `sandbox.toml`, or `production.toml` file to create the final configuration file for the router.

Reviewer: ```suggestion
The `env_specific.toml` file contains values that are specific to the environment. This file is kept separate because the values in it are sensitive and are meant to be set by the user. The `env_specific.toml` file is merged with the `integration_test.toml`, `sandbox.toml`, or `production.toml` file to create the final configuration file for the router.
```

Please ensure you've made the change elsewhere in the document as well.

For example, to build and deploy Hyperswitch in the **sandbox environment**, you can duplicate the `env_specific.toml` file and rename it as `sandbox_config.toml`. Then, update the values in the file with the proper values for the sandbox environment.

The environment-specific `sandbox.toml` file, which contains the Hyperswitch recommended defaults, is merged with the `sandbox_config.toml` file to create the final configuration file called `sandbox_release.toml`. This file is marked as ready for deploying on the sandbox environment.

> 1. Duplicate the `env_specific.toml` file and rename it as `sandbox_config.toml`:
> 
> ```shell
> cp config/deployments/env_specific.toml config/deployments/sandbox_config.toml
> ```

Reviewer: Nit: you can indent the code block so that the code block appears in the same level as the list item, when rendered. Please make this change elsewhere in the document as well.

````suggestion
1. Duplicate the `env_specific.toml` file and rename it as `sandbox_config.toml`:

   ```shell
   cp config/deployments/env_specific.toml config/deployments/sandbox_config.toml
   ```
````

2. Update the values in the `sandbox_config.toml` file with the proper values for the sandbox environment:

```shell
vi config/deployments/sandbox_config.toml
```

3. To merge the files you can use `cat`:

```shell
cat config/deployments/sandbox.toml config/deployments/sandbox_config.toml > config/deployments/sandbox_release.toml
```

> > Note: You can refer to the [`config.example.toml`][config_example] file to understand the variables that used are in the `env_specific.toml` file.

Reviewer: ```suggestion
> [!NOTE]
> You can refer to the [`config.example.toml`][config_example] file to understand the variables that used are in the `env_specific.toml` file.
```

Optional change. This is an "alert" syntax in GitHub flavored Markdown, is rendered slightly better, more info: https://docs.github.com/en/get-started/writing-on-github/getting-started-with-writing-and-formatting-on-github/basic-writing-and-formatting-syntax#alerts

## Scheduler

The scheduler has two components, namely `consumer` and `producer`.

The `consumer.toml` and `producer.toml` files are the configuration files for the `consumer` and `producer`, respectively. These files contain the default values recommended by Hyperswitch.

### Generating a Config File for the Scheduler

Scheduler configuration files are built on top of the router configuration files. So, the `sandbox_release.toml` file is merged with the `consumer.toml` or `producer.toml` file to create the final configuration file for the scheduler.

You can use `cat` to merge the files in the terminal.

- Below is an example for consumer in sandbox environment:

```shell
cat config/deployments/scheduler/consumer.toml config/deployments/sandbox_release.toml > config/deployments/consumer_sandbox_release.toml
```

> Below is an example for producer in sandbox environment:

Reviewer: I'm assuming this was supposed to be an unordered list item?

```suggestion
- Below is an example for producer in sandbox environment:
```

### Generating a Config File for the Router

The `env_specific.toml` file contains values that are specific to the environment. This file is kept separate because the values in it are sensitive and are meant to be set by the user. The `env_specific.toml` file is merged with the `integration_test.toml`, `sandbox.toml`, or `production.toml` file to create the final configuration file for the router.

For example, to build and deploy Hyperswitch in the **sandbox environment**, you can duplicate the `env_specific.toml` file and rename it as `sandbox_config.toml`. Then, update the values in the file with the proper values for the sandbox environment.

The environment-specific `sandbox.toml` file, which contains the Hyperswitch recommended defaults, is merged with the `sandbox_config.toml` file to create the final configuration file called `sandbox_release.toml`. This file is marked as ready for deploying on the sandbox environment.

1. Duplicate the `env_specific.toml` file and rename it as `sandbox_config.toml`:

  ```shell
  cp config/deployments/env_specific.toml config/deployments/sandbox_config.toml

>   ```

Reviewer: Nit: Indent using 3 spaces, not 2.
